### ISSUE TITLE
Add support for directional light masks / Light Shaders

### ISSUE THREAD

**Issue by WrobotGames** [godotengine/godot-proposals#8237]
### Describe the project you are working on

An open world game with a terrain.

### Describe the problem or limitation you are having in your project

When creating a landscape with you often want cloud shadows on your landscape, right now this is not really possible. (The only way to achieve this effect is to have a custom shader on every single mesh, and even then you cant get it to work correctly. )

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

There are 2 solutions: 
1. Implement a light mask property in the directional light settings. Just like the 'projector' property in omni- and spotlights, only with a world wide scaling property. To achieve the effect, you will have to plug in a (tiling) noise texture in here and change the scaling a bit and you have the effect. The texture will work as a multiplier for the light color with world coordinates. Additionally the addition of a fade out distance will make sure the tiling does not become apparent on huge vistas. At that distance the mask will fade to the disabled color. 
2. The complete solution would be to add a new shader type "Light Shaders". This is similar to solution 1, but it enables a lot more possibilities like animated cloud-shadows, or for example cheap water caustics. The shader type would include EMISSION as an output and some kind of UV as input. Light shaders can also be implemented for all other light types.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Solution 1 Mock-up:
![afbeelding](https://github.com/godotengine/godot-proposals/assets/66079814/7798ed9b-8174-41fb-8122-92726d107bd4)
Solution 2 Mock-up:
![afbeelding](https://github.com/godotengine/godot-proposals/assets/66079814/f0bf72e9-f0ae-406e-8755-bb845e025fd5)
The scaling and fade option can probably be implemented in the default light shader.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It would be part of the shaders included in the godot binary.

### Is there a reason why this should be core and not an add-on in the asset library?

The built in shaders would have to be changed.

**Comment by Radivarig** [comment]
Any workarounds for this, something that could be chopped out of other light types that support the projector?
I'd need this for orthogonal directional light to create cloud shadows in one pass and without custom shaders.

**Comment by WrobotGames** [comment]
@Radivarig Depending on the size of the orthographic scene, it should be possible with a spotlight, I think. The biggest downside is that you lose the cascaded shadow map functionality compared to a directional light, and you have to move the light around in code, which might not work very well.  

For scenes with perspective (or big scenes), there is really no option right now, thus this proposal. 

I do not expect someone to pick up this proposal quickly, but once the "shader template" feature gets added, it should be easier to replicate the behaviour of this proposal.

**Comment by Calinou** [comment]
I have an early WIP for cloud shadows here (not functional yet): https://github.com/Calinou/godot/tree/add-cloud-shadows

I don't have the projection code nailed yet, which is the main issue so far. Help is welcome on this area :slightly_smiling_face:

**Comment by QbieShay** [comment]
I mentioned this in rocket chat, I'll add it here for visibility 
1. I think that this can be considered as a feature of world environment rather than the light itself, because I'm questioning what's going to happen if we support multiple projected shadows and multiple directional lights (plus problems with gles possibly)
2. have you considered shader templates as a solution?

**Comment by TheMagnat** [comment]
I started working on this feature, I have a first working implementation, you can check my branch (especially this commit [here](https://github.com/godotengine/godot/commit/d16031023466386987192976dacbec3f8002d1af)).

I've described what I've done and what's remaining to do in the commit description (I've did not mention in but maybe the directionalLight3D in the editor should be have a better UI to not have the projector taking too much space with its parameters).

The worst problem is that the texture repeat badly due to the filtering used. If using mipmaps or linear filtering, since the texture is stored in an atlas, the sides get merged with some empty pixels  (see the third image). I don't know if we could find solution other than creating a sampler array in the shader to send the directional light texture ?

Here is a simple demo project (to try my branch) and some images :

[TestShadow.zip](https://github.com/user-attachments/files/19961942/TestShadow.zip)

![Image](https://github.com/user-attachments/assets/c5cde01e-86e1-4bd1-98e9-ceddbc16c036)
![Image](https://github.com/user-attachments/assets/b6599ee9-34c9-4d0e-b07b-beddd6dfe3c7)
![Image](https://github.com/user-attachments/assets/237974ef-f73b-45a2-84c7-d4a1efc24da3)

**Comment by Jamsers** [comment]
Just wanna chime in to say that if this would work with volumetric fog, it'd now be possible to do gorgeous RDR 2 esque landscape shots in Godot. The current projectors for spotlights and omnilights work with volumetric fog right now, right?

**Comment by Calinou** [comment]
> The current projectors for spotlights and omnilights work with volumetric fog right now, right?

No, not currently: https://github.com/godotengine/godot-proposals/discussions/11987

One difficulty is that volumetric fog only covers a short distance ahead of the camera and is low-frequency data, so it would have trouble representing godrays coming from cloud shadows. Screen-space solutions generally do a better job at this.

**Comment by Mohammad9760** [comment]
light shaders is a really good idea. it will enable use to do so much with lights.
for instance if you want to make a flickering light for a bad florescent bulb or a camp fire, you wouldn't need to animate anything or use a script, just a shader that uses the TIME to sample some noise and apply it to the intensity of the light.
or if you want to make police lights where the lights spin and the blue and red lights fade in and out.
or we could parse IES light profiles and pass the data through uniforms to the light shader to create much more accurate light distributions.
these are just the most obvious use cases I'm sure there's a ton of creative use cases for this that people would benefit from.
is there any plans for adding light shaders?
